### PR TITLE
[macOS] app bundle based packaging

### PR DESCRIPTION
**Pull request by sharvilshah**
Still a bit of a WIP,  but tested locally and it works!

Corresponding PRs on `osquery/osquery` and `codesign` repos shortly. 

For context: we need `.app` bundle so that macOS system can pickup the `embedded.provisionprofile` and we can get ESF events

### CODE REVIEW

**Comment by theopolis** [comment]
Do we need to land this before ES will work?

**Comment by theopolis** [comment]
My preference is to tag a 4.9.0 and test the new packaging and signing workflows. Then we can land this for a planned 5.0.0 release where we collect and merge our breaking changes.

**Comment by mike-myers-tob** [comment]
> My preference is to tag a 4.9.0 and test the new packaging and signing workflows. Then we can land this for a planned 5.0.0 release where we collect and merge our breaking changes.

Ok, if we can't merge this PR though, there's still no working Endpoint Security table in osquery.

**Comment by theopolis** [comment]
Ok, we're ready to merge this when the PR is ready.

**Comment by sharvilshah** [comment]
This is good to go!

File: platform/macos/common.cmake
Comment (theopolis) [inline platform/macos/common.cmake]: Help me understand this section. Are we still installing things into usr/local?

**Review by theopolis — COMMENTED** [review]
Can you provide a preview of what the resulting filesystem layout looks like after this change?

File: platform/macos/common.cmake
Comment (directionless) [reply platform/macos/common.cmake]: I think we should not.

Comment (sharvilshah) [reply platform/macos/common.cmake]: These are the symlinks installed in `/usr/local/bin`:

```
usr
  └── local
      └── bin
          ├── osqueryctl -> /opt/osquery.app/Contents/Resources/osqueryctl
          └── osqueryi -> /opt/osquery.app/Contents/MacOS/osqueryd
```

I am happy to remove these symlinks to `osqueryi` and `osqueryctl`, but I think the consensus was to keep having the  symlink but point to the changed `osqueryd` location, since people might expect `osqueryi` in their `PATH`

**Comment by sharvilshah** [comment]
Here is the resulting filesystem:

```
➜  osquery git:(app_bundle_files) pkgutil --only-files --files io.osquery.agent
opt/osquery.app/Contents/Info.plist
opt/osquery.app/Contents/MacOS/osqueryd
opt/osquery.app/Contents/PkgInfo
opt/osquery.app/Contents/Resources/osqueryctl
opt/osquery.app/Contents/embedded.provisionprofile
private/var/osquery/certs/certs.pem
private/var/osquery/io.osquery.agent.conf
private/var/osquery/io.osquery.agent.plist
private/var/osquery/lenses/access.aug
private/var/osquery/lenses/activemq_conf.aug
private/var/osquery/lenses/activemq_xml.aug
private/var/osquery/lenses/afs_cellalias.aug
private/var/osquery/lenses/aliases.aug
private/var/osquery/lenses/anacron.aug
private/var/osquery/lenses/approx.aug
private/var/osquery/lenses/apt_update_manager.aug
private/var/osquery/lenses/aptcacherngsecurity.aug
private/var/osquery/lenses/aptconf.aug
private/var/osquery/lenses/aptpreferences.aug
private/var/osquery/lenses/aptsources.aug
private/var/osquery/lenses/authorized_keys.aug
private/var/osquery/lenses/automaster.aug
private/var/osquery/lenses/automounter.aug
private/var/osquery/lenses/avahi.aug
private/var/osquery/lenses/backuppchosts.aug
private/var/osquery/lenses/bbhosts.aug
private/var/osquery/lenses/bootconf.aug
private/var/osquery/lenses/build.aug
private/var/osquery/lenses/cachefilesd.aug
private/var/osquery/lenses/carbon.aug
private/var/osquery/lenses/ceph.aug
private/var/osquery/lenses/cgconfig.aug
private/var/osquery/lenses/cgrules.aug
private/var/osquery/lenses/channels.aug
private/var/osquery/lenses/chrony.aug
private/var/osquery/lenses/clamav.aug
private/var/osquery/lenses/cobblermodules.aug
private/var/osquery/lenses/cobblersettings.aug
private/var/osquery/lenses/collectd.aug
private/var/osquery/lenses/cpanel.aug
private/var/osquery/lenses/cron.aug
private/var/osquery/lenses/cron_user.aug
private/var/osquery/lenses/crypttab.aug
private/var/osquery/lenses/csv.aug
private/var/osquery/lenses/cups.aug
private/var/osquery/lenses/cyrus_imapd.aug
private/var/osquery/lenses/darkice.aug
private/var/osquery/lenses/debctrl.aug
private/var/osquery/lenses/desktop.aug
private/var/osquery/lenses/device_map.aug
private/var/osquery/lenses/dhclient.aug
private/var/osquery/lenses/dhcpd.aug
private/var/osquery/lenses/dns_zone.aug
private/var/osquery/lenses/dnsmasq.aug
private/var/osquery/lenses/dovecot.aug
private/var/osquery/lenses/dpkg.aug
private/var/osquery/lenses/dput.aug
private/var/osquery/lenses/erlang.aug
private/var/osquery/lenses/ethers.aug
private/var/osquery/lenses/exports.aug
private/var/osquery/lenses/fai_diskconfig.aug
private/var/osquery/lenses/fonts.aug
private/var/osquery/lenses/fstab.aug
private/var/osquery/lenses/fuse.aug
private/var/osquery/lenses/gdm.aug
private/var/osquery/lenses/getcap.aug
private/var/osquery/lenses/group.aug
private/var/osquery/lenses/grub.aug
private/var/osquery/lenses/gshadow.aug
private/var/osquery/lenses/gtkbookmarks.aug
private/var/osquery/lenses/host_conf.aug
private/var/osquery/lenses/hostname.aug
private/var/osquery/lenses/hosts.aug
private/var/osquery/lenses/hosts_access.aug
private/var/osquery/lenses/htpasswd.aug
private/var/osquery/lenses/httpd.aug
private/var/osquery/lenses/inetd.aug
private/var/osquery/lenses/inifile.aug
private/var/osquery/lenses/inittab.aug
private/var/osquery/lenses/inputrc.aug
private/var/osquery/lenses/interfaces.aug
private/var/osquery/lenses/iproute2.aug
private/var/osquery/lenses/iptables.aug
private/var/osquery/lenses/iscsid.aug
private/var/osquery/lenses/jaas.aug
private/var/osquery/lenses/jettyrealm.aug
private/var/osquery/lenses/jmxaccess.aug
private/var/osquery/lenses/jmxpassword.aug
private/var/osquery/lenses/json.aug
private/var/osquery/lenses/kdump.aug
private/var/osquery/lenses/keepalived.aug
private/var/osquery/lenses/known_hosts.aug
private/var/osquery/lenses/koji.aug
private/var/osquery/lenses/krb5.aug
private/var/osquery/lenses/ldif.aug
private/var/osquery/lenses/ldso.aug
private/var/osquery/lenses/lightdm.aug
private/var/osquery/lenses/limits.aug
private/var/osquery/lenses/login_defs.aug
private/var/osquery/lenses/logrotate.aug
private/var/osquery/lenses/logwatch.aug
private/var/osquery/lenses/lokkit.aug
private/var/osquery/lenses/lvm.aug
private/var/osquery/lenses/mailscanner.aug
private/var/osquery/lenses/mailscanner_rules.aug
private/var/osquery/lenses/masterpasswd.aug
private/var/osquery/lenses/mcollective.aug
private/var/osquery/lenses/mdadm_conf.aug
private/var/osquery/lenses/memcached.aug
private/var/osquery/lenses/mke2fs.aug
private/var/osquery/lenses/modprobe.aug
private/var/osquery/lenses/modules.aug
private/var/osquery/lenses/modules_conf.aug
private/var/osquery/lenses/mongodbserver.aug
private/var/osquery/lenses/monit.aug
private/var/osquery/lenses/multipath.aug
private/var/osquery/lenses/mysql.aug
private/var/osquery/lenses/nagioscfg.aug
private/var/osquery/lenses/nagiosobjects.aug
private/var/osquery/lenses/netmasks.aug
private/var/osquery/lenses/networkmanager.aug
private/var/osquery/lenses/networks.aug
private/var/osquery/lenses/nginx.aug
private/var/osquery/lenses/nrpe.aug
private/var/osquery/lenses/nslcd.aug
private/var/osquery/lenses/nsswitch.aug
private/var/osquery/lenses/ntp.aug
private/var/osquery/lenses/ntpd.aug
private/var/osquery/lenses/odbc.aug
private/var/osquery/lenses/opendkim.aug
private/var/osquery/lenses/openshift_config.aug
private/var/osquery/lenses/openshift_http.aug
private/var/osquery/lenses/openshift_quickstarts.aug
private/var/osquery/lenses/openvpn.aug
private/var/osquery/lenses/oz.aug
private/var/osquery/lenses/pagekite.aug
private/var/osquery/lenses/pam.aug
private/var/osquery/lenses/pamconf.aug
private/var/osquery/lenses/passwd.aug
private/var/osquery/lenses/pbuilder.aug
private/var/osquery/lenses/pg_hba.aug
private/var/osquery/lenses/pgbouncer.aug
private/var/osquery/lenses/php.aug
private/var/osquery/lenses/phpvars.aug
private/var/osquery/lenses/postfix_access.aug
private/var/osquery/lenses/postfix_main.aug
private/var/osquery/lenses/postfix_master.aug
private/var/osquery/lenses/postfix_passwordmap.aug
private/var/osquery/lenses/postfix_sasl_smtpd.aug
private/var/osquery/lenses/postfix_transport.aug
private/var/osquery/lenses/postfix_virtual.aug
private/var/osquery/lenses/postgresql.aug
private/var/osquery/lenses/properties.aug
private/var/osquery/lenses/protocols.aug
private/var/osquery/lenses/puppet.aug
private/var/osquery/lenses/puppet_auth.aug
private/var/osquery/lenses/puppetfile.aug
private/var/osquery/lenses/puppetfileserver.aug
private/var/osquery/lenses/pylonspaste.aug
private/var/osquery/lenses/pythonpaste.aug
private/var/osquery/lenses/qpid.aug
private/var/osquery/lenses/quote.aug
private/var/osquery/lenses/rabbitmq.aug
private/var/osquery/lenses/radicale.aug
private/var/osquery/lenses/redis.aug
private/var/osquery/lenses/reprepro_uploaders.aug
private/var/osquery/lenses/resolv.aug
private/var/osquery/lenses/rhsm.aug
private/var/osquery/lenses/rmt.aug
private/var/osquery/lenses/rsyncd.aug
private/var/osquery/lenses/rsyslog.aug
private/var/osquery/lenses/rtadvd.aug
private/var/osquery/lenses/rx.aug
private/var/osquery/lenses/samba.aug
private/var/osquery/lenses/schroot.aug
private/var/osquery/lenses/securetty.aug
private/var/osquery/lenses/sep.aug
private/var/osquery/lenses/services.aug
private/var/osquery/lenses/shadow.aug
private/var/osquery/lenses/shells.aug
private/var/osquery/lenses/shellvars.aug
private/var/osquery/lenses/shellvars_list.aug
private/var/osquery/lenses/simplelines.aug
private/var/osquery/lenses/simplevars.aug
private/var/osquery/lenses/sip_conf.aug
private/var/osquery/lenses/slapd.aug
private/var/osquery/lenses/smbusers.aug
private/var/osquery/lenses/solaris_system.aug
private/var/osquery/lenses/soma.aug
private/var/osquery/lenses/spacevars.aug
private/var/osquery/lenses/splunk.aug
private/var/osquery/lenses/squid.aug
private/var/osquery/lenses/ssh.aug
private/var/osquery/lenses/sshd.aug
private/var/osquery/lenses/sssd.aug
private/var/osquery/lenses/star.aug
private/var/osquery/lenses/stunnel.aug
private/var/osquery/lenses/subversion.aug
private/var/osquery/lenses/sudoers.aug
private/var/osquery/lenses/sysconfig.aug
private/var/osquery/lenses/sysconfig_route.aug
private/var/osquery/lenses/sysctl.aug
private/var/osquery/lenses/syslog.aug
private/var/osquery/lenses/systemd.aug
private/var/osquery/lenses/termcap.aug
private/var/osquery/lenses/thttpd.aug
private/var/osquery/lenses/tmpfiles.aug
private/var/osquery/lenses/trapperkeeper.aug
private/var/osquery/lenses/tuned.aug
private/var/osquery/lenses/up2date.aug
private/var/osquery/lenses/updatedb.aug
private/var/osquery/lenses/util.aug
private/var/osquery/lenses/vfstab.aug
private/var/osquery/lenses/vmware_config.aug
private/var/osquery/lenses/vsftpd.aug
private/var/osquery/lenses/webmin.aug
private/var/osquery/lenses/wine.aug
private/var/osquery/lenses/xendconfsxp.aug
private/var/osquery/lenses/xinetd.aug
private/var/osquery/lenses/xml.aug
private/var/osquery/lenses/xorg.aug
private/var/osquery/lenses/xymon.aug
private/var/osquery/lenses/xymon_alerting.aug
private/var/osquery/lenses/yaml.aug
private/var/osquery/lenses/yum.aug
private/var/osquery/osquery.example.conf
private/var/osquery/packs/hardware-monitoring.conf
private/var/osquery/packs/incident-response.conf
private/var/osquery/packs/it-compliance.conf
private/var/osquery/packs/osquery-monitoring.conf
private/var/osquery/packs/ossec-rootkit.conf
private/var/osquery/packs/osx-attacks.conf
private/var/osquery/packs/unwanted-chrome-extensions.conf
private/var/osquery/packs/vuln-management.conf
private/var/osquery/packs/windows-attacks.conf
private/var/osquery/packs/windows-hardening.conf
```

Screen Shot 2021-07-02 at 9 33 58 AM

File: platform/macos/common.cmake
Comment (theopolis) [reply platform/macos/common.cmake]: Let's keep it for this PR and follow up with a second if we do want to remove these (I think we do). The approach I recommend on macOS is to start installing a `/etc/paths.d/osquery` file with our `/opt/osquery.app/...` path.

**Review by theopolis — DISMISSED** [review]

**Comment by directionless** [comment]
Do we think `/var/osquery` is correct? Or should that land in `/opt/osquery/var`? (I honestly have no idea)

File: platform/macos/common.cmake
Comment (directionless) [inline platform/macos/common.cmake]: I think this needs to change

File: platform/macos/common.cmake
Comment (directionless) [inline platform/macos/common.cmake]: same

File: platform/macos/common.cmake
Comment (directionless) [inline platform/macos/common.cmake]: `/opt/bin` ?

**Review by directionless — CHANGES_REQUESTED** [review]
Paths look wrong.

**Comment by sharvilshah** [comment]
@directionless Yup paths need to be updated to what we talked about. Addressing them in the next PR, I just want to see if CI passes with this merged in (builds do work locally)!

File: platform/macos/common.cmake
Comment (sharvilshah) [reply platform/macos/common.cmake]: This is just the `package_data` holding place

**Comment by theopolis** [comment]
I also think we should pin osquery/osquery to a packaging commit so that we can merge things here and then be in sync with what changes are needed in the osquery repo.

**Comment by theopolis** [comment]
To test in either direction, we can set the CI checkout to point to a fork with the changes.

**Comment by theopolis** [comment]
I'd like to land https://github.com/osquery/osquery/pull/7208 before merging this change.

**Comment by theopolis** [comment]
@directionless and @sharvilshah, I tried to update the paths here based on your conversation from office hours. Please let me know if I understood the plan correctly.
- Use `/opt/osquery` as the new "home"
- Use `/opt/osquery/lib/osquery.app` for the macos-specific bundle
- Continue to install symlinks for binaries into `/usr/local/bin`

**Review by directionless — APPROVED** [review]
Let's try!

**Pull request merged**
